File: Tests/Tests/STPImageLibraryTest.m
//
//  STPImageLibraryTest.m
//  Stripe
//
//  Created by Jack Flintermann on 6/30/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <OCMock/OCMock.h>





#import "STPTestUtils.h"

@interface STPImageLibraryTest : XCTestCase
@property id mockLocale;
@property NSArray<NSNumber *> *cardBrands;
@end

@implementation STPImageLibraryTest

- (void)setUp {
    [super setUp];

    self.mockLocale = OCMClassMock([NSLocale class]);
    OCMStub([self.mockLocale currentLocale]).andReturn(self.mockLocale);
    OCMStub([self.mockLocale localeIdentifier]).andReturn(@"EN_US");

    self.cardBrands = @[
                        @(STPCardBrandAmex),
                        @(STPCardBrandDinersClub),
                        @(STPCardBrandDiscover),
                        @(STPCardBrandJCB),
                        @(STPCardBrandMastercard),
                        @(STPCardBrandUnionPay),
                        @(STPCardBrandUnknown),
                        @(STPCardBrandVisa),
                        ];
}

- (void)tearDown {
    [super tearDown];

    [self.mockLocale stopMocking];
}

- (void)testCardIconMethods {
    STPAssertEqualImages([STPImageLibrary applePayCardImage], [STPImageLibrary safeImageNamed:@"stp_card_applepay" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary amexCardImage], [STPImageLibrary safeImageNamed:@"stp_card_amex" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary dinersClubCardImage], [STPImageLibrary safeImageNamed:@"stp_card_diners" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary discoverCardImage], [STPImageLibrary safeImageNamed:@"stp_card_discover" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary jcbCardImage], [STPImageLibrary safeImageNamed:@"stp_card_jcb" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary mastercardCardImage], [STPImageLibrary safeImageNamed:@"stp_card_mastercard" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary unionPayCardImage], [STPImageLibrary safeImageNamed:@"stp_card_unionpay_en" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary visaCardImage], [STPImageLibrary safeImageNamed:@"stp_card_visa" templateIfAvailable:NO]);
    STPAssertEqualImages([STPImageLibrary unknownCardCardImage], [STPImageLibrary safeImageNamed:@"stp_card_unknown" templateIfAvailable:YES]);
}

- (void)testBrandImageForCardBrand {
    for (NSNumber *brandNumber in self.cardBrands) {
        STPCardBrand brand = (STPCardBrand)[brandNumber integerValue];
        UIImage *image = [STPImageLibrary brandImageForCardBrand:brand];

        switch (brand) {
            case STPCardBrandVisa:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_visa" templateIfAvailable:NO]);
                break;
            case STPCardBrandAmex:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_amex" templateIfAvailable:NO]);
                break;
            case STPCardBrandMastercard:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_mastercard" templateIfAvailable:NO]);
                break;
            case STPCardBrandDiscover:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_discover" templateIfAvailable:NO]);
                break;
            case STPCardBrandJCB:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_jcb" templateIfAvailable:NO]);
                break;
            case STPCardBrandDinersClub:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_diners" templateIfAvailable:NO]);
                break;
            case STPCardBrandUnionPay:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unionpay_en" templateIfAvailable:NO]);
                break;
            case STPCardBrandUnknown:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unknown" templateIfAvailable:YES]);
                break;
        }
    }
}

- (void)testBrandImageForCardBrand_zh {
    [self.mockLocale stopMocking];

    self.mockLocale = OCMClassMock([NSLocale class]);
    OCMStub([self.mockLocale currentLocale]).andReturn(self.mockLocale);
    OCMStub([self.mockLocale localeIdentifier]).andReturn(@"ZH_HANT");

    UIImage *image = [STPImageLibrary brandImageForCardBrand:STPCardBrandUnionPay];
    STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unionpay_zh" templateIfAvailable:NO]);
}

- (void)testTemplatedBrandImageForCardBrand {
    for (NSNumber *brandNumber in self.cardBrands) {
        STPCardBrand brand = (STPCardBrand)[brandNumber integerValue];
        UIImage *image = [STPImageLibrary templatedBrandImageForCardBrand:brand];

        switch (brand) {
            case STPCardBrandVisa:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_visa_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandAmex:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_amex_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandMastercard:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_mastercard_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandDiscover:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_discover_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandJCB:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_jcb_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandDinersClub:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_diners_template" templateIfAvailable:YES]);
                break;
            case STPCardBrandUnionPay:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unionpay_template_en" templateIfAvailable:YES]);
                break;
            case STPCardBrandUnknown:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unknown" templateIfAvailable:YES]);
                break;
        }
    }
}

- (void)testTemplatedBrandImageForCardBrand_zh {
    [self.mockLocale stopMocking];

    self.mockLocale = OCMClassMock([NSLocale class]);
    OCMStub([self.mockLocale currentLocale]).andReturn(self.mockLocale);
    OCMStub([self.mockLocale localeIdentifier]).andReturn(@"ZH_HANT");

    UIImage *image = [STPImageLibrary templatedBrandImageForCardBrand:STPCardBrandUnionPay];
    STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_unionpay_template_zh" templateIfAvailable:YES]);
}

- (void)testCVCImageForCardBrand {
    for (NSNumber *brandNumber in self.cardBrands) {
        STPCardBrand brand = (STPCardBrand)[brandNumber integerValue];
        UIImage *image = [STPImageLibrary cvcImageForCardBrand:brand];

        switch (brand) {
            case STPCardBrandAmex:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_cvc_amex" templateIfAvailable:NO]);
                break;

            default:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_cvc" templateIfAvailable:NO]);
                break;
        }
    }
}

- (void)testErrorImageForCardBrand {
    for (NSNumber *brandNumber in self.cardBrands) {
        STPCardBrand brand = (STPCardBrand)[brandNumber integerValue];
        UIImage *image = [STPImageLibrary errorImageForCardBrand:brand];

        switch (brand) {
            case STPCardBrandAmex:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_error_amex" templateIfAvailable:NO]);
                break;

            default:
                STPAssertEqualImages(image, [STPImageLibrary safeImageNamed:@"stp_card_error" templateIfAvailable:NO]);
                break;
        }
    }
}

@end

File: Stripe/StripeiOSTests/STPImageLibraryTest.swift
//
//  STPImageLibraryTest.swift
//  StripeiOS Tests
//
//  Created by Cameron Sabol on 4/7/22.
//  Copyright © 2022 Stripe, Inc. All rights reserved.
//

import XCTest

@testable@_spi(STP) import Stripe
@testable@_spi(STP) import StripeCore
@testable@_spi(STP) import StripePayments
@testable@_spi(STP) import StripePaymentSheet
@testable@_spi(STP) import StripePaymentsUI
@testable@_spi(STP) import StripeUICore

class STPImageLibraryTestSwift: XCTestCase {

    static let cardBrands: [STPCardBrand] = [
        .amex,
        .cartesBancaires,
        .dinersClub,
        .discover,
        .JCB,
        .mastercard,
        .unionPay,
        .unknown,
        .visa,
    ]

    func testCardIconMethods() {
        STPAssertEqualImages(
            STPImageLibrary.amexCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_amex", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.dinersClubCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_diners", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.discoverCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_discover", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.jcbCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_jcb", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.mastercardCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_mastercard", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.unionPayCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_unionpay", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.visaCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_visa", templateIfAvailable: false)
        )
        STPAssertEqualImages(
            STPImageLibrary.unknownCardCardImage(),
            STPImageLibrary.safeImageNamed("stp_card_unknown", templateIfAvailable: false)
        )
    }

    func testBrandImageForCardBrand() {
        for brand in Self.cardBrands {
            let image = STPImageLibrary.brandImage(for: brand, template: false)

            switch brand {
            case .visa:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_visa", templateIfAvailable: false)
                )
            case .amex:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_amex", templateIfAvailable: false)
                )
            case .mastercard:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed(
                        "stp_card_mastercard",
                        templateIfAvailable: false
                    )
                )
            case .discover:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_discover", templateIfAvailable: false)
                )
            case .JCB:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_jcb", templateIfAvailable: false)
                )
            case .dinersClub:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_diners", templateIfAvailable: false)
                )
            case .unionPay:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_unionpay", templateIfAvailable: false)
                )
            case .cartesBancaires:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_cartes_bancaires", templateIfAvailable: false)
                )
            case .unknown:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_unknown", templateIfAvailable: false)
                )
            }
        }
    }

    func testUnpaddedImageForCardBrands() {
        for brand in STPCardBrand.allCases {
            let image = STPImageLibrary.unpaddedCardBrandImage(for: brand)
            // Assert image exists
            XCTAssert(image.size != .zero)
        }
    }

    func testCVCImageForCardBrand() {
        for brand in Self.cardBrands {
            let image = STPImageLibrary.cvcImage(for: brand)

            switch brand {
            case .amex:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_cvc_amex", templateIfAvailable: false)
                )
            default:
                STPAssertEqualImages(
                    image,
                    STPImageLibrary.safeImageNamed("stp_card_cvc", templateIfAvailable: false)
                )
            }
        }
    }

    func testErrorImageForCardBrand() {
        for brand in Self.cardBrands {
            let image = STPImageLibrary.errorImage(for: brand)
            STPAssertEqualImages(
                image,
                STPImageLibrary.safeImageNamed("stp_card_error", templateIfAvailable: false)
            )
        }
    }

    func testMiscImages() {
        STPAssertEqualImages(
            STPImageLibrary.bankIcon(),
            STPImageLibrary.safeImageNamed("stp_icon_bank", templateIfAvailable: false)
        )
    }

    func testBankIconCodeImagesExist() {
        for iconCode in PaymentSheetImageLibrary.BankIconCodeRegexes.keys {
            XCTAssertNotNil(
                PaymentSheetImageLibrary.bankIcon(for: iconCode),
                "Missing image for \(iconCode)"
            )
        }
    }

    func testBankNameToIconCode() {
        // bank of america
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "bank of america"), "boa")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "BANK of AMERICA"), "boa")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "BANKof AMERICA"), "default")

        // capital one
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "capital one"), "capitalone")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Capital One"), "capitalone")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Capital      One"), "default")

        // citibank
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "citibank"), "citibank")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Citibank"), "citibank")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Citi Bank"), "default")

        // compass
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "bbva"), "compass")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "BBVA"), "compass")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "compass"), "compass")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "b b v a"), "default")

        // morganchase
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Morgan Chase"), "morganchase")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "morgan chase"), "morganchase")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "jp morgan"), "morganchase")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "JP Morgan"), "morganchase")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Chase"), "morganchase")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "chase"), "morganchase")

        // pnc
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "pncbank"), "pnc")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "PNCBANK"), "pnc")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "pnc bank"), "pnc")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "PNC Bank"), "pnc")

        // suntrust
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "suntrust"), "suntrust")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "SUNTRUST"), "suntrust")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "suntrust bank"), "suntrust")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Suntrust Bank"), "suntrust")

        // svb
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Silicon Valley Bank"), "svb")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "SILICON VALLEY BANK"), "svb")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "SILICONVALLEYBANK"), "default")

        // usaa
        XCTAssertEqual(
            PaymentSheetImageLibrary.bankIconCode(for: "USAA Federal Savings Bank"),
            "usaa"
        )
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "USAA Bank"), "usaa")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "USAA Savings Bank"), "default")

        // usbank
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "US Bank"), "usbank")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "U.S. Bank"), "usbank")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "u.s. Bank"), "usbank")

        // wellsfargo
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Wells Fargo"), "wellsfargo")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "WELLS FARGO"), "wellsfargo")
        XCTAssertEqual(PaymentSheetImageLibrary.bankIconCode(for: "Well's Fargo"), "default")
    }

}
